### DISCUSSION TITLE
[Question/Suggestion] Partial record access with RecordDotSyntax

### DISCUSSION THREAD

**Discussion by googleson78** [ghc-proposals/ghc-proposals#459] [General]
(I also posted this to the cafe, but after I learned about this place, I think here might be a better place for this)

With the advent of RecordDotSyntax (or more precisely OverloadedRecordDot?) in 9.2, I was expecting this "wart" to be fixed:

```haskell
data Ty
  = X {x :: String}
  | Y {y :: Int}

> y $ X "lol"
*** Exception: No match in record selector y
```

As it seems like the perfect opportunity to introduce a "breaking change" for a behaviour that is usually undesirable (in my experience).

But instead:
```haskell
> :set -XOverloadedRecordDot
> (X "lol").y
*** Exception: No match in record selector y
```

Unfortunately, I couldn't follow and/or remember all of the discussion around RecordDotSyntax, so this might be something I missed.

Is there a reason why we don't generate `HasField "x" r a` only when all of the constructors for `r` have a field `"x" :: a`?

If not, is the community open to changing this, while the new extension is still ripe?

To clarify, I'm not suggesting that we change the partial accessors that ghc usually generates (we can already sort of "fix" that with `NoFieldSelectors`). I'm suggesting that we don't carry over this behaviour to `RecordDotSyntax`.

**Comment by adamgundry** [comment]
A difficulty is that `HasField` should (ideally) be operationally the same as the existing selector function. Otherwise compile-time performance suffers due to having to compile additional functions for each field, of which there can be many.

I've been slowly working on a proposal to redesign `HasField`, primarily to tackle the story around record update, but another thing I'd like to address is precisely this issue. (See https://github.com/adamgundry/ghc-proposals/blob/hasfield-redesign/proposals/0000-hasfield-redesign.rst#partial-fields for the relevant part of the current unfinished draft. Help welcome!)

My current suggestion (not yet set in stone) is that we introduce a new warning for uses of incomplete record selectors (either directly or via `HasField`). One could then compile with `-Werror=incomplete-record-selectors` to forbid all such uses. This fits with other places in GHC where partiality is allowed but warned about (cf. https://github.com/ghc-proposals/ghc-proposals/pull/351). However it wouldn't allow a choice to inhibit the selector to be made at a datatype definition site that affects all use sites, and perhaps there's a case for that.

Another possibility would be to restrict the solving of `HasField` constraints like you suggest, i.e. GHC wouldn't automatically solve a `HasField` constraint for partial fields, but perhaps the user could provide their own instance (perhaps even treating the field as if it had a `Maybe` type). But I'm not sure if we would want to make that behaviour uniform, or if there is a sensible way to control it per-module or per-definition.

**Reply by googleson78** [reply]
If always generating the functions is really going to be an issue for compiler perf(I guess actually implementing and benchmarking this is the only way to check..), then I'd also be very happy with an extension to turn this behaviour on together with the warning you're suggesting, preferably on by default.

**Reply by adamgundry** [reply]
I've previously benchmarked the nontrivial impact of generating update functions (as envisaged by the accepted proposal for `setField`): https://gitlab.haskell.org/ghc/ghc/-/merge_requests/3257#note_311465 That said, perhaps some variants of selectors might not be as bad, because the definitions are simpler, and perhaps we can do better (e.g. by not always generating the functions up front, or by optimizing the patch or GHC elsewhere).

If what you actually want is for GHC not to solve (or issue a warning/error when solving) `HasField` constraints where the field is partial, that doesn't require changing the operational behaviour of selectors so the performance issue doesn't arise. It would arise if we wanted such `HasField` constraints to automatically be solved with a `Maybe` field type, for example.

**Reply by AntC2** [reply]
> @adamgundry,  `HasField` should (ideally) be operationally the same as the existing selector function. 

Consider this variant of o.p.'s datatype

```haskell
    data Ty2 where                                  -- GADT syntax, but not a GADT
              X2 :: a ~ String => { x2 :: a } -> Ty2 
              Y2 :: b ~ Int    => { y2 :: b } -> Ty2

    yy2 = case blah2 of
            -- Y2 {} -> case (y2 blah2) of _ -> "ignored"
                                  -- rejected `cannot use record selector `y2` as a function because of escaped type variables`
                                  --          `Probable fix: use pattern-matching syntax instead`
            Y2 { y2 } -> y2       -- OK, accepted: but don't the tyvars escape just as much?
         -- X2 { x2 } -> x2       -- if uncommented, rejected: mismatch of result types

```

My thought/hope was that partial field record selectors should be valid only inside the pattern match. Something like a local `HasField` constraint exposed inside a GADT. But if I go on to that:

```haskell
    data Ty3 a where                                -- properly a GADT
                X3 :: a ~ String => { x3 :: a } -> Ty3 a
                Y3 :: b ~ Int    => { y3 :: b } -> Ty3 b

    yy3 = case blah3 of
            Y3 {} -> y3 blah3  -- record selector happy: why is this unlike Y2 ?
                               -- aren't tyvars escaping just as much ?
         -- X3 {} -> x3 blah3  -- hmmm if I uncomment this, accepted but [-Winaccessible-code], [-Woverlapping-patterns]
```

(Exercised at GHC 8.10.2, by the way. Also accepted at 7.10.3, except if the `X3 {} -> ` line uncommented rejects on mismatch of result types -- as expected.)

**Reply by adamgundry** [reply]
Hmm, I think there are two separate issues here:

1. For selection with a "partial" field, assuming we had `-Wincomplete-record-selectors`, it would be nice if GHC could suppress the warning when the field is statically known to be present. For example we could plausibly accept these without a warning:
```hs
data Ty
  = X {x :: String}
  | Y {y :: Int}

eg1 = x $ X "lol"
eg2 t = case t of
    X{} -> t.x
    Y{} -> ""
```

Perhaps we could reuse the existing pattern-match checker infrastructure to establish when an application of a selector function must be to one of the constructors that has the field. The details would probably require a bit of work though.

2. Existential type variables prevent record selector functions being generated. Your `Ty2` has existential type variables (syntactically) even though the equality constraints make it essentially equivalent to `Ty`. Morally the selector function should have a type like `x2 :: Ty2 -> exists a . a ~ String * a` but that requires first-class existential types, so GHC simply doesn't produce a selector in this case. This isn't the case with `Ty3` because its type variables are not existential, so we get `x3 :: Ty3 a -> a`. I can see that this is annoying if we know statically which constructor is being used for a particular selection, and hence could in principle avoid the existential, but I think that will be harder to resolve and probably isn't worth it.

**Comment by AntC2** [comment]
> ... a behaviour that is usually undesirable (in my experience).
> 
> ..., I'm not suggesting that we change the partial accessors that ghc usually generates (we can already sort of "fix" that with `NoFieldSelectors`). I'm suggesting that we don't carry over this behaviour to `RecordDotSyntax`.

Presumably people still want to access those other fields by field name(?) So how should they do that? A pattern match with field name punning? 

```haskell
case blah  of
    X { x } -> ...
    Y { y } -> ...
```

- a) That can just as much throw a no-match exception;
- b) Now readers of the code have to follow two different methods to access fields (and beware `x, y` would be variables rather than methods/functions);
- c) So you might as well pattern-match everywhere, not use `RecordDotSyntax`, not use `FieldSelectors`.

The "wart" is really in the form of the `data` declaration. I don't think that tinkering at the selecting end will out that damned spot.

**Reply by tomjaguarpaw** [reply]
> Presumably people still want to access those other fields by field name(?) So how should they do that? A pattern match with field name punning?
>
> a) That can just as much throw a no-match exception;

Can it? How? (I can by omitting a constructor but that's a different issue.)

**Reply by googleson78** [reply]
> Presumably people still want to access those other fields by field name(?) So how should they do that?

Write an accessor that returns `Maybe`, or push in the direction that ghc can generate "virtual" fields returning `Maybe`? Otherwise, I don't see how you could access something that isn't there, without checking if it actually is (pattern matching).

> a) That can just as much throw a no-match exception;

I don't see how from "there are other ways to make a mistake" it follows that "we shouldn't remove ways to make a mistake".

Additionally, I also don't understand how that throws a `no-match` exception:
```haskell
{-# LANGUAGE NamedFieldPuns #-}

data T
  = X {x :: Char}
  | Y {y :: Int}

f :: T -> String
f x =
  case x of
    X{y} -> show y
    Y{x} -> show x
```

This doesn't compile:
```haskell
    • Constructor ‘X’ does not have field ‘y’
    • In the pattern: X {y}
      In a case alternative: X {y} -> show y
      In the expression:
        case x of
          X {y} -> show y
          Y {x} -> show x
   |
10 |     X{y} -> show y
   |       ^
Failed, no modules loaded.
```

Can you provide an example?

> b) Now readers of the code have to follow two different methods to access fields (and beware x, y would be variables rather than methods/functions);

Which are they? The only way to access this code with my suggestion is with pattern matching, since you don't have accessors/fields, since not all the constructors have the field you're trying to access.

**Reply by AntC2** [reply]
>> a) That can just as much throw a no-match exception;

> Can it? How? (I can by omitting a constructor but that's a different issue.)

> I don't see how from "there are other ways to make a mistake" it follows that "we shouldn't remove ways to make a mistake".
>
> Additionally, I also don't understand how that throws a no-match exception:

I mean, in general code using pattern-matching might give a no-match. (Ok you might ask for a Warning.) As far as I can see it's exactly comparable: the equivalent to `blah.y` is

>        case blah  of
>            -- X { x } -> ...       -- `blah.y` wouldn't generate this branch
>            Y { y } -> ...

> Write an accessor that returns `Maybe`, ...

So now there's _three_ styles of the field name I have to bear in mind reading your code; and none of those for partial fields "remove ways to make a mistake":

- always-present fields use a selector _function_ (possibly via `OverloadedRecordDot`).
- pattern-matched fields use punning so the field name appears as a _variable_.
- hand-coded accessor functions use a `Maybe`. So the field name appears in a `fromJust` (?) -- which can of course throw a no-match exception.
- Oh, and some code might branch on data constructor, using `OverloadedRecordDot` safely within the branch. Are you wanting to ban that style? Or do you want this suggestion to be aware of that?

```haskell
case blah of
  Y {} -> ... blah.y ...          -- `OverloadedRecordDot` is safe here
```

All this suggestion is achieving is a different shade of lipstick on the pig. I'm agreeing with @adamgundry " `HasField` should (ideally) be operationally the same as the existing selector function." -- for me not because of efficiency concerns, but because of mental overload concerns.

**Comment by adamgundry** [comment]
I've made a proposal for my suggestion for fixing this issue: https://github.com/ghc-proposals/ghc-proposals/pull/516
